tooling/miniutil/src/build/terminator.rs

```diff
@@ -1,5 +1,255 @@
+        self.finish_block(expose_provenance(dest, ptr, next_block.name_into_u32()));
+        self.set_block(next_block)
+    }
+
+    pub fn with_exposed_provenance(&mut self, dest: PlaceExpr, addr: ValueExpr) {
+        let next_block = self.declare_block();
+        self.finish_block(with_exposed_provenance(dest, addr, next_block.name_into_u32()));
+        self.set_block(next_block)
+    }
+
+    pub fn lock_create(&mut self, ret: PlaceExpr) {
+        let next_block = self.declare_block();
+        self.finish_block(lock_create(ret, next_block.name_into_u32()));
+        self.set_block(next_block)
+    }
+
+    pub fn lock_acquire(&mut self, lock_id: ValueExpr) {
+        let next_block = self.declare_block();
+        self.finish_block(lock_acquire(lock_id, next_block.name_into_u32()));
+        self.set_block(next_block)
+    }
+
+    pub fn lock_release(&mut self, lock_id: ValueExpr) {
+        let next_block = self.declare_block();
+        self.finish_block(lock_release(lock_id, next_block.name_into_u32()));
+        self.set_block(next_block)
+    }
+
+    // terminators with 2 or more following blocks
+    pub fn if_<F, G>(&mut self, condition: ValueExpr, then_branch: F, else_branch: G)
+    where
+        F: Fn(&mut Self),
+        G: Fn(&mut Self),
+    {
+        self.switch_int(bool_to_int::<u8>(condition), &[(1, &then_branch)], else_branch);
+    }
+
+    pub fn switch_int<T, G>(
+        &mut self,
+        value: ValueExpr,
+        cases: &[(T, &dyn Fn(&mut Self))],
+        fallback: G,
+    ) where
+        T: Clone + Into<Int>,
+        G: Fn(&mut Self),
+    {
+        // closures + blocks we we run the closures on
+        let mut branches: Vec<(&dyn Fn(&mut Self), CurBlock)> = Vec::new();
+        // branch map for switch terminator
+        let mut branch_map: Map<Int, BbName> = Map::new();
+
+        for (case, branch) in cases {
+            let new_block = self.declare_block();
+            branch_map.try_insert(case.clone().into(), new_block.name).unwrap();
+            branches.push((branch, new_block));
+        }
+
+        let fallback_block = self.declare_block();
+        let switch = Terminator::Switch { value, cases: branch_map, fallback: fallback_block.name };
+        self.finish_block(switch);
+
+        // None means that every branch finished on its own and we don't need a next block
+        let mut after_switch_block: Option<CurBlock> = None;
+
+        // for the execution the fallback can be handled as any other branch
+        branches.push((&fallback, fallback_block));
+
+        for (branch, block) in branches {
+            self.set_block(block);
+            branch(self);
+            // if current block not finished jump to next_block
+            if self.cur_block.is_some() {
+                if after_switch_block.is_none() {
+                    after_switch_block = Some(self.declare_block());
+                }
+                self.goto(after_switch_block.as_ref().unwrap().name);
```

Comment:
A more idiomatic way to do this, without `unwrap`, would be something like
```rust
let jump_to_block = after_switch_block.get_or_insert_with(|| self.declare_block());
self.goto(jump_to_block);
```

Reply:
Thank you for this advice. Was struggling here a bit in how to do that in Rust. :)